src/spoke/Spoke.sol

```diff
@@ -299,42 +299,50 @@ abstract contract Spoke is ISpoke, Multicall, NoncesKeyed, AccessManagedUpgradea
   ) external onlyPositionManager(onBehalfOf) returns (uint256, uint256) {
     Reserve storage reserve = _getReserve(reserveId);
     UserPosition storage userPosition = _userPositions[onBehalfOf][reserveId];
+    PositionStatus storage positionStatus = _positionStatus[onBehalfOf];
     _validateRepay(reserve);
 
+    IHubBase.PremiumDelta memory premiumDelta;
+    uint256 drawnDebtRestored = amount;
     uint256 drawnIndex = reserve.hub.getAssetDrawnIndex(reserve.assetId);
-    (uint256 drawnDebtRestored, , uint256 premiumDebtRayRestored) = _getUserDebt(
+    (drawnDebtRestored, premiumDelta.restoredPremiumRay) = _calculateRestoreAmount(
       userPosition,
-      drawnIndex
+      drawnIndex,
+      drawnDebtRestored
     );
 
-    (drawnDebtRestored, premiumDebtRayRestored) = _calculateRestoreAmount(
-      drawnDebtRestored,
-      premiumDebtRayRestored,
-      amount
-    );
+    // replicate hub.previewRestoreByAssets() instead of calling here
+    // verified later against actual restored shares returned from hub.restore()
+    uint256 previewRestoredShares = drawnDebtRestored.rayDivDown(drawnIndex);
+    userPosition.drawnShares -= previewRestoredShares.toUint120();
 
-    IHubBase.PremiumDelta memory premiumDelta = userPosition.getPremiumDelta({
+    premiumDelta = userPosition.getPremiumDelta({
       drawnIndex: drawnIndex,
-      riskPremium: 0,
-      restoredPremiumRay: premiumDebtRayRestored
+      riskPremium: positionStatus.riskPremium,
+      restoredPremiumRay: premiumDelta.restoredPremiumRay
     });
 
     uint256 totalDebtRestored = drawnDebtRestored + premiumDelta.restoredPremiumRay.fromRayUp();
     reserve.underlying.safeTransferFrom(msg.sender, address(reserve.hub), totalDebtRestored);
-    uint256 restoredShares = reserve.hub.restore(reserve.assetId, drawnDebtRestored, premiumDelta);
+    require(
+      previewRestoredShares == reserve.hub.restore(reserve.assetId, drawnDebtRestored, premiumDelta)
+    );
 
     userPosition.applyPremiumDelta(premiumDelta);
```

Comment:
let's apply premium delta before restore, since we update drawnShares before as well